### PR TITLE
Multithreaded S3 file upload

### PR DESCRIPTION
**Pull request by konrad-c**
This update changes the push_app functionality to upload files to S3 in a parallel fashion. This code has reduced the deploy time of a project from 35 minutes to 6 minutes when using 10 threads and even when using a conservative 5 threads, a build time of 10 minutes.
There are issues associated with the unit tests which expect file yields which are not yielded with an asynchronous, multi-threaded function.
Additionally other tests which expect uploaded file list to match an array. Since order cannot be guaranteed in multi-threaded functions this may require changes. 
The build does not pass due to the tests incorrectly handling multi-threading. Some collaboration on getting the tests working is needed and welcomed.
Travis CI build link: https://travis-ci.org/konrad-c/dpl/jobs/351088693

### CODE REVIEW

**Comment by mboudreau** [comment]
This is the issue for this: https://github.com/travis-ci/dpl/issues/765

By using threads, it borked a lot of the tests surrounding push_app and we're both newbs when it comes to ruby and I'm not certain on how we can add tests that work with threads.  I was hoping people in Travis can help us with this (@BanzaiMan).

Thanks.

**Comment by BanzaiMan** [comment]
I can take a look at fixing the specs. Do you have a build log using this code? (You can email us at support@travis-ci.com if you'd prefer that.)

**Comment by mboudreau** [comment]
@BanzaiMan Sent via email :)

**Comment by mboudreau** [comment]
@BanzaiMan Any updates?

**Comment by BanzaiMan** [comment]
@mboudreau Sorry about the delay. I am planning to spend time on fixing this in the next few days.

**Comment by mboudreau** [comment]
Great, thanks :D

**Comment by BanzaiMan** [comment]
I've worked out fixes for the spec, but for some reason I cannot open a PR for @konrad-c's fork. I've contacted GitHub support to see what's up. If this doesn't get resolved, I'll open a separate PR to supersede this one.

**Comment by mboudreau** [comment]
@BanzaiMan Awesome news!

So, you can create your own personal fork from the repo, or create a new branch in the repo based on this PR, do your modifications, commit/push, then create a [PR on konrad's fork by going here](https://github.com/travis-ci/dpl/compare/master...konrad-c:master) and specifying the fork/branch you want to PR with.

**Comment by BanzaiMan** [comment]
Ah, I didn't think of that. https://github.com/konrad-c/dpl/pull/1 should do it.

**Comment by konrad-c** [comment]
@BanzaiMan Build was successful. Thanks again for all your help in rewriting the tests!

File: lib/dpl/provider/s3.rb [hunk 1]
```diff
@@ -43,21 +43,11 @@ def upload_path(filename)
...
-            opts[:server_side_encryption] = "AES256" if options[:server_side_encryption]
-            unless File.directory?(filename)
-              log "uploading #{filename.inspect} with #{opts.inspect}"
-              result = api.bucket(option(:bucket)).object(upload_path(filename)).upload_file(filename, opts)
-              warn "error while uploading #{filename.inspect}" unless result
-            end
-          end
+          files = Dir.glob(*glob_args)
         end
+        upload_multithreaded(files)
```
Comment (BanzaiMan) [inline lib/dpl/provider/s3.rb]: We should select things which are not directories.

```ruby
upload_multithreaded(file.reject {|f| File.directory?(f)}
```

File: lib/dpl/provider/s3.rb [hunk 2]
```diff
@@ -70,6 +60,39 @@ def push_app
         end
       end
 
+      def upload_multithreaded(files, thread_count = 5)
+        file_number = 0
+        mutex = Mutex.new
+        threads = []
+        log "Beginning upload of #{files.length} files with #{thread_count} threads."
```
Comment (BanzaiMan) [inline lib/dpl/provider/s3.rb]: … so that `file.length` here is correct.

**Comment by mboudreau** [comment]
@BanzaiMan thanks for the suggestions, makes sense :)

Is there a particular reason to use Dir.chdir insteaf of just doing a glob pattern with the directory prefix?

**Comment by BanzaiMan** [comment]
I don't think there is a reason; just personal taste. (It does not matter in this case, IMO.)

**Comment by konrad-c** [comment]
@BanzaiMan Changed the push_app function to not use the chdir and those warnings are gone. Changed the tests as well to reflect this.

**Comment by BanzaiMan** [comment]
@konrad-c Sweet! Thanks for the PR!

**Comment by BanzaiMan** [comment]
I plan to release this early next week.

**Pull request merged**